### DISCUSSION TITLE
Skip faulty lines

### DISCUSSION THREAD

**Discussion by Felienne** [hedyorg/hedy#3192] [Archive]
In a talk I saw about [Gidget](http://www.helpgidget.org/) (very cool and inspiring in any case!) I noticed they do something supersmart... they simply skip faulty lines, accompanied with an error messages saying the line will be skipped.

Since we have meta enabled in the whole transpiler (since #3155) we can probably do something similar, but how to communicate this to kids is a bit unclear, so I am making this into a discussion rather than an issue.

**Comment by Felienne** [comment]
The more I think about this, the more I really like this idea, esp since we also have the UI and the mechanics to indicate skipped lines with the "sleepy dude":

<img width="333" alt="image" src="https://user-images.githubusercontent.com/1003685/202491243-261579e6-4a44-4a26-81e3-f5aa4e67ab3f.png">

Would love to see this happening as a student project!

**Comment by Felienne** [comment]
Great news! @ToniSkulj would love to pick this up as his BSc project!

**Comment by ToniSkulj** [comment]
I think Gidget removed this feature, notice that Gidget is not back at his original place! Which he should be if the last _left_ were to be executed.


![image](https://user-images.githubusercontent.com/48225550/220311967-8e5a79a1-6ddd-4fb3-8498-0af05ef5ce68.png)

**Comment by ToniSkulj** [comment]
A little update.

The plan is to skip faulty code (no matter where this line is and how many there are), currently, when a user adds faulty code a ParseError is raised. We would need to replace this with a rule _parse_error_. Essentially a rule that will match anything after all rules have been tried.

In the visitor method of this rule, the faulty code would be replaced with nothing (an empty string). We would still need to know where this faulty code is and what it is so we can give feedback to the user. This is where https://github.com/hedyorg/hedy/pull/4094 comes into play. 

After finishing/merging https://github.com/hedyorg/hedy/pull/4094 we can start implementing this. Below is a working example (run locally):

Hedy code:
```
name is ask 'what is your name?'
if name is Hedy print 'nice' else print 'boo!'
prind "Wrong use of print"
```

Notice that the last print statement is wrong. The first 2 lines will execute but the last one is removed,
visible in the Sourcemapper output:
```
3/80-3/106 --- prind "Wrong use of print":
        None/None-None/None ---

```

After that, we could extend this idea to other errors.

**Comment by ToniSkulj** [comment]
The sourcemapper has been merged  #4094 🥳 
Currently, I'm working on skipping faulty code, progress can be found in the PR #4320
